src/Product.jl

```diff
@@ -0,0 +1,94 @@
+@doc doc"""
+    Product{TM<:Tuple, TRanges<:Tuple, TSizes<:Tuple} <: Manifold
+
+Product manifold $M_1 \times M_2 \times \dots \times M_n$ with product geometry.
+`TRanges` and `TSizes` statically define the relationship between representation
+of the product manifold and representations of point, tangent vectors
+and cotangent vectors of respective manifolds.
+
+# Constructor
+
+    Product(M_1, M_2, ..., M_n)
+
+generates the product manifold $M_1 \times M_2 \times \dots \times M_n$
+"""
+struct Product{TM<:Tuple, TRanges, TSizes} <: Manifold
+    manifolds::TM
+end
+
+function Product(manifolds...)
+    sizes = map(m -> representation_size(m, MPoint), manifolds)
+    lengths = map(s -> prod(s), sizes)
+    ranges = UnitRange{Int64}[]
+    k = 1
+    for len ∈ lengths
+        push!(ranges, k:(k+len-1))
+        k += len
+    end
+    TRanges = tuple(ranges...)
+    TSizes = tuple(sizes...)
+    return Product{typeof(manifolds), TRanges, TSizes}(manifolds)
+end
+
+function representation_size(M::Product, ::Type{T}) where {T}
+    return (sum(map(m -> representation_size(m, T), M.manifolds)),)
```

Comment:
As below, this could be
```julia
function representation_size(M::Product, ::Type{T}) where {T}
    return representation_size(first(M), T) + representation_size(tail(M), T)
end
```

Reply:
That would work if `M.manifolds` only contains two manifolds, but it could contain an arbitrary number.

Reply:
Could do a `mapreduce`.

Reply:
> That would work if `M.manifolds` only contains two manifolds, but it could contain an arbitrary number.

Actually the second part recursively upwraps it :) Notice that `tail` in the second summand works on all remaining one (again doing the first plus the remaining ones and so on)... nice recursion!

Reply:
Oh! Cool, nice approach.

Reply:
Result of `tail` would have to be re-wrapped into a new object of type `Product` which doesn't look ideal. `mapreduce` makes sense.

Reply:
Or we would at least need a new function that operates on tuples of manifolds.